### PR TITLE
[feat] Evaluate and log all performance values before asserting them

### PR DESCRIPTION
**Pull request by mand35**
One can create multiple Objectives for a singe test case (see example below). This commit improves the logging of the values. It add the tag to the log and it logs all values. (before it was all up to reference is not reached). Thus we can keep better track of all the values observed during the tests. 

Reference example:
       p_names = {'creation', 'stat', 'removal'}
        for p_name in p_names:
           self.perf_patterns[p_name] = sn.extractsingle(r'^\s+File '+p_name+
                    '\s+:\s+(?P<'+p_name+'>\S+) ', self.stdout, p_name, float)

        self.reference = {
            'kupe:compute' : {
                'creation' : (7747,  -(2*223.1)/7747, None),
                'stat' :     (16527, -(2*558.2)/16527,None),
                'removal' :  (7355, -(2*173.1)/7355, None)
            },
        }

### CODE REVIEW

**Comment by vkarak** [comment]
@mand35 Thanks for your PR. The performance logging is completely redesigned in #213, which is planned to be merged this week. There you can specify exactly the format of the performance log in the configuration file and you can log the performance variables (tags) as well.

File: reframe/core/pipeline.py [hunk 1]
```diff
@@ -1039,21 +1039,27 @@ def check_performance(self):
...
                     raise SanityError(
                         "tag `%s' not resolved in references for `%s'" %
                         (tag, self._current_partition.fullname)
                     )
-                evaluate(assert_reference(value, ref, low_thres, high_thres))
+            for tag, expr in self.perf_patterns.items():
+                key = '%s:%s' % (self._current_partition.fullname, tag)
+                ref, low_thres, high_thres = self.reference[key]
+                evaluate(assert_reference(value[tag], ref, low_thres,
+                         high_thres))
```
Comment (vkarak) [inline reframe/core/pipeline.py]: I can't understand very well the purpose of this nested loop here.

Comment (mand35) [reply reframe/core/pipeline.py]: It is not a nested loop. I needed to move the evaluation to an additional loop. The reason was that if one of the early reference comparison throw the assert the other values were not written to the log file. In this way first all values are written and then the comparison is performed.

**Comment by vkarak** [comment]
@jenkins-cscs retry all

File: reframe/core/pipeline.py (hunk 1 above)
Comment (vkarak) [reply reframe/core/pipeline.py]: You are right, for some reason I overlooked the alignment! I see what you are trying to do.

File: reframe/core/pipeline.py [hunk 2]
```diff
@@ -1039,21 +1039,27 @@ def check_performance(self):
             return
 
         with os_ext.change_dir(self._stagedir):
+            # first check and print all values with references
+            value = {}
```
Comment (vkarak) [inline reframe/core/pipeline.py]: Better call this `perf_values`. I'd also change a bit the comment to "Evaluate all performance variables first, so as to log them, then assert them".

File: reframe/core/pipeline.py [hunk 3]
```diff
@@ -1039,21 +1039,27 @@ def check_performance(self):
...
+                        (tag, value[tag], self.reference[key])
                     )
+                    ref, low_thres, high_thres = self.reference[key]
                 except KeyError:
                     raise SanityError(
                         "tag `%s' not resolved in references for `%s'" %
                         (tag, self._current_partition.fullname)
                     )
-                evaluate(assert_reference(value, ref, low_thres, high_thres))
+            for tag, expr in self.perf_patterns.items():
```
Comment (vkarak) [inline reframe/core/pipeline.py]: There is no need to go over the `perf_patterns` again here. Just over the evaluated `perf_values`.

Also, leave a blank line after the previous `for` loop.

File: reframe/core/pipeline.py [hunk 4]
```diff
@@ -1039,21 +1039,27 @@ def check_performance(self):
...
+                value[tag] = evaluate(expr)
                 key = '%s:%s' % (self._current_partition.fullname, tag)
                 try:
-                    ref, low_thres, high_thres = self.reference[key]
                     self._perf_logger.info(
-                        'value: %s, reference: %s' %
-                        (value, self.reference[key])
+                        '%s, value: %s, reference: %s' %
+                        (tag, value[tag], self.reference[key])
                     )
```
Comment (vkarak) [inline reframe/core/pipeline.py]: For the logging, I suggest waiting for #213 to be merged (perhaps today).

**Comment by vkarak** [comment]
Closing this. Replaced by #359.

**Pull request closed without merging**